### ISSUE TITLE
Statuer sur l'identification d'un utilisateur

### ISSUE THREAD

**Issue by GerardPaligot** [ChantyTaguan/zds-site#19]
Aujourd'hui, et pour de bonnes raisons, nous avons choisi `Profile` pour identifier un utilisateur dans les nouveaux modèles des notifications et des souscriptions. L'idée de base est bonne et correspond bien avec ce que nous voulons faire à l'avenir puisque, dans un avenir proche j'espère, nous utiliserons partout `Profile`.

Cependant, à l'heure d'aujourd'hui, c'est `User` qui est utilisé partout et ça entraine une grande complexité dans le code puisque, finalement, on a beaucoup plus souvent de `User` que de `Profile`.

L'objectif de cette issue est donc de savoir si nous gardons `Profile` comme identification ou si nous switchons sur `User`. Si nous switchons sur `User`, je veux bien m'en charger ce week-end (ce soir, c'est Koh Lanta :-°). Si nous gardons `Profile`,et comme `User` est utilisé dans les autres modèles, partout où nous récupérons un `user` à partir d'un `profile`, nous devons vérifier qu'il possède bien un `user` (puisque ce n'est pas le cas pour certains utilisateurs très particulier).

**Comment by ChantyTaguan** [comment]
C'est délicat. D'un côté, si on finit effectivement par migrer sur `Profile` partout, autant garder `Profile` ici, ce sera ça de moins à migrer. D'un autre, si c'est pour pourrir le code pour faire du workaround provisoire qu'il faudra virer une fois tout passé sur `Profile`, ça vaut peut-être pas la peine de chipoter....

Question : il me semble que ce n'est pas normal que certains `User`existent sans `Profile`. Est-ce que ce ne serait pas plus logique de corriger plutôt ça ? Où est-ce que ça arrive ? Uniquement dans certains tests ? Y a-t-il une raison ?

**Comment by GerardPaligot** [comment]
Tout à fait. C'est une situation un peu délicate puisque si nous revenons à `User`, dans un futur plus ou moins proche nous allons devoir repasser à `Profile`. Et dans l'état actuel, des choses, il faut gérer les utilisateurs sans profil donc pourrir le code.

> Question : il me semble que ce n'est pas normal que certains Userexistent sans Profile. Est-ce que ce ne serait pas plus logique de corriger plutôt ça ? Où est-ce que ça arrive ? Uniquement dans certains tests ? Y a-t-il une raison ?

S'il existe une raison, un bug ou autre, je n'en sais rien. En fait, je ne sais même pas quand et à quel moment ces membres sont créés. D'ailleurs, si ça se trouve, le bug n'est pas en prod. Donc oui, c'est peut-être plus intéressant de corriger ce bug mais j'ai encore aucune idée de savoir comment faire. Par contre, si nous le corrigeons, je pense que nous devons directement contribuer sur le dépôt officiel pour rebase ici.

**Comment by GerardPaligot** [comment]
C'est ça, en fait ce sont à cause des fixtures de test qui créés des `User` et non pas des `Profile`.

**Comment by GerardPaligot** [comment]
Nouvelle info : Un utilisateur anonyme n'a forcément pas de profil.

**Comment by ChantyTaguan** [comment]
C'est-à-dire ? C'est l'absence de profile qui fait qu'on sait que c'est un anonyme ? C'est bizarre... On peut pas changer ça ? De toute façon ça sera changé quand on passera tout sur `Profile` aussi, non ? Sauf si ça demande un gros refactoring quoi. Mais si c'est pas méga complexe, autant en profiter maintenant et faire une PR qui change ça sur dev. Après, on rebase, et on peut garder les `Profile` ici.

**Comment by GerardPaligot** [comment]
> C'est l'absence de profile qui fait qu'on sait que c'est un anonyme ?

Non, `user.is_anonymous()`.

> De toute façon ça sera changé quand on passera tout sur `Profile` aussi, non ?

Pour les utilisateurs anonymes non.

> Sauf si ça demande un gros refactoring quoi. Mais si c'est pas méga complexe, autant en profiter maintenant et faire une PR qui change ça sur dev. Après, on rebase, et on peut garder les Profile ici.

Le merge de `Profile` et `User` est un problème complexe sur lequel SpaceFox s'est déjà penché et a déjà fait une [ébauche](https://github.com/SpaceFox/zds-site/tree/fix_2711_user_profile).

**Comment by GerardPaligot** [comment]
Malgré ma dernière information, je vote pour rester sur du `Profile` finalement. Je pense que ça va nous faire chier toi et moi de changer vers `User` puis de rechanger vers `Profile` et finalement, le bug identifié n'est spécifique qu'à une mauvaise utilisation des tests.

Que penses-tu de confirmer notre choix sur le `Profile` donc ?
